File: CPP/abc204_e.cpp
#include <iostream>
#include <cmath>
#include <queue>
#include <vector>

#include <boost/range/irange.hpp>

using namespace std;
using namespace boost;

struct Vertex;

struct Connection {
	Vertex *target;
	long c, d, minTime;

	Connection(Vertex *target, long c, long d) : target(target), c(c), d(d) {
//		long t = (long)sqrt(d);
//		minTime = (t > 1 && calcTime(t - 1) < calcTime(t)) ? t - 1 : t;
		minTime = (long)round(sqrt(d) - 1);
	}

	long calcTime(long time) const { return time + c + d / (time + 1); }
};

struct Vertex {
	vector<Connection> connections;
	long distance = -1;
};

struct SearchNode {
	Vertex *target;
	long time;
};

bool operator <(const SearchNode &a, const SearchNode &b) { return a.time > b.time; }

int main() {
	long N, M;
	cin >> N >> M;

	Vertex vertices[N];

	for (long i : irange(M)) {
		long a, b, c, d;
		cin >> a >> b >> c >> d;
		a--; b--;

		vertices[a].connections.emplace_back(&vertices[b], c, d);
		vertices[b].connections.emplace_back(&vertices[a], c, d);
	}

	Vertex *goal = &vertices[N - 1];

	priority_queue<SearchNode> queue;

	queue.push({&vertices[0], 0L});

	while (!queue.empty()) {
		auto [target, time] = queue.top();
		queue.pop();

		if (target->distance >= 0) continue;
		target->distance = time;

		if (target == goal) break;

		for (const Connection &connection : target->connections) {
			if (connection.target->distance >= 0) continue;
			long t = max(connection.minTime, time);
			queue.push({connection.target, connection.calcTime(t)});
		}
	}

	cout << goal->distance << endl;

	return 0;
}

File: Kotlin/abc204_e.kt
import java.util.*
import kotlin.collections.ArrayList
import kotlin.math.*

class Connection(val target: Vertex, val c: Long, val d: Long) {
	val minTime = round(sqrt(d.toDouble()) - 1).toLong()

	fun calcTime(time: Long) = time + c + d / (time + 1)
}

class Vertex {
	val connections = ArrayList<Connection>()
	var distance = -1L
}

class SearchNode(val target: Vertex, val time: Long) : Comparable<SearchNode> {
	override fun compareTo(other: SearchNode) = (time - other.time).sign
}

fun main() {
	val input = Scanner(System.`in`)
	val N = input.nextInt()
	val M = input.nextInt()

	val vertices = Array(N) { Vertex() }

	for (i in 0 until M) {
		val a = input.nextInt() - 1
		val b = input.nextInt() - 1
		val c = input.nextLong()
		val d = input.nextLong()
		vertices[a].connections.add(Connection(vertices[b], c, d))
		vertices[b].connections.add(Connection(vertices[a], c, d))
	}

	val goal = vertices.last()

	val queue = PriorityQueue<SearchNode>()
	queue.add(SearchNode(vertices[0], 0))

	while (queue.isNotEmpty()) {
		val node = queue.poll()

		if (node.target.distance >= 0) continue
		node.target.distance = node.time

		if (node.target == goal) break

		for (connection in node.target.connections) {
			if (connection.target.distance >= 0) continue
			val t = max(connection.minTime, node.time)
			queue.add(SearchNode(connection.target, connection.calcTime(t)))
		}
	}

	println(goal.distance)
}
